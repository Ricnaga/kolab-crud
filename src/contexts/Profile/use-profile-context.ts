import { useUpdateUserMutation, useUserQuery } from "@/infra/users";
import useContext from "@/shared/hooks/useContext";
import {
  ChangeEvent,
  createContext,
  FormEvent,
  useEffect,
  useState,
} from "react";

export enum FormFields {
  name = "Nome",
  username = "Usuário",
  email = "E-mail",
  addressStreet = "Rua",
  addressSuite = "Complemento",
  addressCity = "Cidade",
  addressZipcode = "Cep",
  phone = "Telefone",
  website = "Website",
  companyName = "Empresa",
}

type FieldState = Record<keyof typeof FormFields, string>;

export function useProfileProvider() {
  const { data, isLoading } = useUserQuery({ id: "1" });
  const { mutate, isPending } = useUpdateUserMutation();

  const [isEnabled, setAsEnabled] = useState<boolean>(false);
  const [fields, setFields] = useState<FieldState>({
    name: "",
    username: "",
    email: "",
    addressStreet: "",
    addressSuite: "",
    addressCity: "",
    addressZipcode: "",
    phone: "",
    website: "",
    companyName: "",
  });

  useEffect(() => {
    setFields({
      name: data?.name || "",
      username: data?.username || "",
      email: data?.email || "",
      addressStreet: data?.address.street || "",
      addressSuite: data?.address.suite || "",
      addressCity: data?.address.city || "",
      addressZipcode: data?.address.zipcode || "",
      phone: data?.phone || "",
      website: data?.website || "",
      companyName: data?.company.name || "",
    });
  }, [data]);

  const handleChange = (e: ChangeEvent<HTMLInputElement>, key: string) =>
    setFields({ ...fields, [key]: e.target.value });

  function onSubmit(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();
    mutate(
      {
        id: data?.id || 0,
        name: fields.name,
        username: fields.username,
        email: fields.email,
        phone: fields.phone,
        website: fields.website,
        company: {
          name: fields.companyName,
          catchPhrase: data?.company.catchPhrase || "",
          bs: data?.company.bs || "",
        },
        address: {
          street: fields.addressStreet,
          suite: fields.addressSuite,
          city: fields.addressCity,
          zipcode: fields.addressZipcode,
          geo: {
            lat: data?.address.geo.lat || "",
            lng: data?.address.geo.lng || "",
          },
        },
      },
      {
        onSuccess: () => console.log("ATUALIZADO COM SUCESSO"),
        onError: () => console.error("ERRO AO ATUALIZAR COMENTÁRIO"),
      }
    );
  }

  return {
    isEnabled,
    setAsEnabled,
    fields,
    setFields,
    isLoading,
    isPending,
    onSubmit,
    handleChange,
  };
}

export type UseProfileProviderReturnType = ReturnType<
  typeof useProfileProvider
>;

export const ProfileContext = createContext({} as UseProfileProviderReturnType);

function useProfileContext() {
  return useContext({
    context: ProfileContext,
    hookName: useProfileContext.name,
    providerName: "User",
  });
}

export default useProfileContext;
