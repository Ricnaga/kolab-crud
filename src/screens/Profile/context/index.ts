import useProfileContext, {
  UseProfileProviderReturnType,
  useProfileProvider,
} from "./use-profile-context";

export {
  useProfileContext as default,
  type UseProfileProviderReturnType,
  useProfileProvider,
};
