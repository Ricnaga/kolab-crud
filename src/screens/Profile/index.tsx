import { useUpdateUserMutation, useUserQuery } from "@/infra/users";
import {
  Button,
  CardRoot,
  Checkbox,
  Field,
  Input,
  SimpleGrid,
  Skeleton,
  Spinner,
} from "@chakra-ui/react";
import { FormEvent, useEffect, useState } from "react";

type FieldState = {
  name: string;
  username: string;
  email: string;
  addressStreet: string;
  addressSuite: string;
  addressCity: string;
  addressZipcode: string;
  phone: string;
  website: string;
  companyName: string;
};

export default function Profile() {
  const { data, isLoading } = useUserQuery({ id: "1" });
  const { mutate, isPending } = useUpdateUserMutation();

  const [isEnabled, setAsEnabled] = useState<boolean>(false);
  const [fields, setFields] = useState<FieldState>({
    name: "",
    username: "",
    email: "",
    addressStreet: "",
    addressSuite: "",
    addressCity: "",
    addressZipcode: "",
    phone: "",
    website: "",
    companyName: "",
  });

  useEffect(() => {
    setFields({
      name: data?.name || "",
      username: data?.username || "",
      email: data?.email || "",
      addressStreet: data?.address.street || "",
      addressSuite: data?.address.suite || "",
      addressCity: data?.address.city || "",
      addressZipcode: data?.address.zipcode || "",
      phone: data?.phone || "",
      website: data?.website || "",
      companyName: data?.company.name || "",
    });
  }, [data]);

  if (isLoading)
    return (
      <CardRoot
        display={"grid"}
        p={4}
        gap={8}
        mt={2}
        maxW={538}
        mx={"auto"}
        marginTop={10}
      >
        {Array.from({ length: 10 }).map(() => (
          <Skeleton key={Math.random().toString()} h={10} />
        ))}
      </CardRoot>
    );

  function handleUpdate(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();
    mutate(
      {
        id: data?.id || 0,
        name: fields.name,
        username: fields.username,
        email: fields.email,
        phone: fields.phone,
        website: fields.website,
        company: {
          name: fields.companyName,
          catchPhrase: data?.company.catchPhrase || "",
          bs: data?.company.bs || "",
        },
        address: {
          street: fields.addressStreet,
          suite: fields.addressSuite,
          city: fields.addressCity,
          zipcode: fields.addressZipcode,
          geo: {
            lat: data?.address.geo.lat || "",
            lng: data?.address.geo.lng || "",
          },
        },
      },
      {
        onSuccess: () => console.log("ATUALIZADO COM SUCESSO"),
        onError: () => console.error("ERRO AO ATUALIZAR COMENTÁRIO"),
      }
    );
  }

  return (
    <form onSubmit={handleUpdate}>
      <SimpleGrid
        gap={2}
        w={"full"}
        h={"full"}
        maxW={"32rem"}
        mx={"auto"}
        marginTop={"1rem"}
      >
        <Field.Root>
          <Field.Label>Nome</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) => setFields({ ...fields, name: e.target.value })}
            value={fields.name}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Usuário</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) => setFields({ ...fields, username: e.target.value })}
            value={fields.username}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>E-mail</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) => setFields({ ...fields, email: e.target.value })}
            value={fields.email}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Rua</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) =>
              setFields({ ...fields, addressStreet: e.target.value })
            }
            value={fields.addressStreet}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Complemento</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) =>
              setFields({ ...fields, addressSuite: e.target.value })
            }
            value={fields.addressSuite}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Cidade</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) =>
              setFields({ ...fields, addressCity: e.target.value })
            }
            value={fields.addressCity}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>CEP</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) =>
              setFields({ ...fields, addressZipcode: e.target.value })
            }
            value={fields.addressZipcode}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Telefone</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) => setFields({ ...fields, phone: e.target.value })}
            value={fields.phone}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Website</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) => setFields({ ...fields, website: e.target.value })}
            value={fields.website}
          />
        </Field.Root>
        <Field.Root>
          <Field.Label>Empresa</Field.Label>
          <Input
            disabled={!isEnabled}
            bg={"yellow.50"}
            px={2}
            className="peer"
            placeholder=""
            onChange={(e) =>
              setFields({ ...fields, companyName: e.target.value })
            }
            value={fields.companyName}
          />
        </Field.Root>
        <Checkbox.Root
          variant={"subtle"}
          onCheckedChange={(e) => setAsEnabled(!!e.checked)}
          checked={isEnabled}
        >
          <Checkbox.HiddenInput />
          <Checkbox.Control>
            <Checkbox.Indicator />
          </Checkbox.Control>
          <Checkbox.Label>Habilitar campos</Checkbox.Label>
        </Checkbox.Root>
        <Button
          disabled={!isEnabled || isPending}
          type="submit"
          colorPalette={"cyan"}
        >
          {isPending ? <Spinner size={"sm"} /> : "Alterar"}
        </Button>
      </SimpleGrid>
    </form>
  );
}
