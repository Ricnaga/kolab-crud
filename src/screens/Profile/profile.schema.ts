import { zodResolver } from "@hookform/resolvers/zod";
import z from "zod";

const required_error = "Esse campo é obrigatório";

export const profileSchema = z
  .object({
    name: z.string({ required_error }),
    username: z.string({ required_error }),
    email: z.string({ required_error }),
    addressStreet: z.string({ required_error }),
    addressSuite: z.string({ required_error }),
    addressCity: z.string({ required_error }),
    addressZipcode: z.string({ required_error }),
    phone: z.string({ required_error }),
    website: z.string({ required_error }),
    companyName: z.string({ required_error }),
  })
  .required();

export const profileResolver = zodResolver(profileSchema);

export type ProfileSchemaType = z.infer<typeof profileSchema>;
