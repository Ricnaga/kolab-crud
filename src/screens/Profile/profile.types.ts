import { UseProfileReturnType } from "./use-profile.model";

export type ProfileViewProps = UseProfileReturnType;
