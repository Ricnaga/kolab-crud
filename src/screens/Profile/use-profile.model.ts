import { useForm } from "react-hook-form";
import { profileResolver, ProfileSchemaType } from "./profile.schema";
import { useAppSelector } from "@/contexts/RTK/store";
import { userValue } from "@/contexts/RTK/features/user/user.slice";
import { useState } from "react";
import { useUpdateUserMutation } from "@/infra/users";

export const useProfile = () => {
  const user = useAppSelector(userValue);

  const { isPending, mutate } = useUpdateUserMutation();

  const { register, handleSubmit } = useForm<ProfileSchemaType>({
    resolver: profileResolver,
    values: {
      ...user,
      addressStreet: user.address.street,
      addressSuite: user.address.suite,
      addressCity: user.address.city,
      addressZipcode: user.address.zipcode,
      companyName: user.company.name,
    },
  });

  const [isEnabled, setAsEnabled] = useState<boolean>(false);

  const onSubmit = handleSubmit((data) =>
    mutate(
      {
        id: Math.round(Math.random() * 10),
        name: data.name,
        username: data.username,
        email: data.email,
        phone: data.phone,
        website: data.website,
        company: {
          ...user.company,
          name: data.companyName,
        },
        address: {
          street: data.addressStreet,
          suite: data.addressSuite,
          city: data.addressCity,
          zipcode: data.addressZipcode,
          geo: user.address.geo,
        },
      },
      {
        onSuccess: () => console.log("ATUALIZADO COM SUCESSO"),
        onError: () => console.error("ERRO AO ATUALIZAR COMENT√ÅRIO"),
      }
    )
  );

  return { register, onSubmit, isEnabled, setAsEnabled, isPending };
};

export type UseProfileReturnType = ReturnType<typeof useProfile>;
