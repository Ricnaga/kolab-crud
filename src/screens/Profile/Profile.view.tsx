import { Button, SimpleGrid, Spinner } from "@chakra-ui/react";
import { ProfileViewProps } from "./profile.types";
import { FormFields } from "@/contexts/Profile/use-profile-context";
import { InputProfile } from "./components/InputProfile";
import { CheckboxProfile } from "./components/CheckboxProfile";

export const ProfileView = (props: ProfileViewProps) => {
  const { isEnabled, isPending, onSubmit, register, setAsEnabled } = props;

  return (
    <form onSubmit={onSubmit}>
      <SimpleGrid
        gap={2}
        w="full"
        h="full"
        maxW="32rem"
        justifySelf="center"
        marginTop="1rem"
      >
        {Object.entries(FormFields).map(([key, value]) => (
          <InputProfile
            key={key}
            label={value}
            {...register(key as keyof typeof FormFields, {
              disabled: !isEnabled,
            })}
          />
        ))}

        <CheckboxProfile
          onCheckedChange={(e) => setAsEnabled(!!e.checked)}
          checked={isEnabled}
          label="Habilitar campos"
        />
        <Button
          disabled={!isEnabled || isPending}
          type="submit"
          colorPalette={"cyan"}
        >
          {isPending ? <Spinner size={"sm"} /> : "Alterar"}
        </Button>
      </SimpleGrid>
    </form>
  );
};
