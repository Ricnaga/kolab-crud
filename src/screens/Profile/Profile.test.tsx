import { describe, expect, it, renderWithTheme } from "@vitest-utils";
import Profile from ".";

describe("Page: Profile", () => {
  it("should render correctly", () => {
    const { container } = renderWithTheme(<Profile />);

    expect(container).toMatchSnapshot();
  });
});
