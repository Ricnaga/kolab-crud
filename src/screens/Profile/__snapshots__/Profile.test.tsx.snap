// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page: Profile > should render correctly 1`] = `
<div>
  <div
    class="css-1f9j1y1"
  >
    <div
      class="chakra-card__root css-1v1qa66"
    >
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
      <div
        class="chakra-skeleton css-ht60kb"
      />
    </div>
  </div>
</div>
`;
